crates/ruff/src/rules/ruff/typing.rs

```diff
@@ -135,7 +101,12 @@ impl<'a> TypingTarget<'a> {
                     )
                 }
             }
-            Expr::BinOp(..) => Some(TypingTarget::PEP604Union(expr)),
```

Comment:
This was the key issue: every BinOp was getting turned into `PEP604Union`, and when we went to analyze that expression in `PEP604UnionIterator`, we'd see that it wasn't a `BitOr`, and return that expression immediately, then analyze it again here, etc.

Reply:
Oh, nice catch! I wonder what's the use case of any other operator in this context (will look at the linked project to satisfy my curiosity).